FF145 Storage Access HTTP headers (#41614)
mdn/content: files/en-us/web/api/storage_access_api/index.md

Since there is no privacy risk, because it is not possible to track users across sites via partitioned cookies, browsers send partitioned cookies in requests and make them available to the embedded resources.
Note however that, because the cookies aren't shared between sites, they are also not automatically synchronized across sites.
Browsers have various mechanisms to partition third-party cookie access, for example [Firefox Total Cookie Protection](https://blog.mozilla.org/en/mozilla/firefox-rolls-out-total-cookie-protection-by-default-to-all-users-worldwide/) and [Cookies Having Independent Partitioned State (CHIPS)](/en-US/docs/Web/Privacy/Guides/Privacy_sandbox/Partitioned_cookies).

When we talk about third-party cookies in the context of the Storage Access API, we implicitly mean _unpartitioned_ third-party cookies.

### How it works

Third party content embedded in an {{htmlelement("iframe")}} that needs to access cookie or other unpartitioned state can request access using the Storage Access API as follows:

> 1. {{domxref("Document.hasStorageAccess()")}} can be called to check if it already has access to unpartitioned cookies.

Reviewer: ```suggestion
1. {{domxref("Document.hasStorageAccess()")}} can be called to check if the embedded content already has access to unpartitioned cookies.
```

   - Firefox only prompts users after an origin has requested storage access on more than a threshold number of sites.
   - Chrome shows prompts for all embedded content that has not previously received storage access.
     It will however automatically grant access and skip prompts if the embedded content and embedding site are part of the same [related website set](/en-US/docs/Web/API/Storage_Access_API/Related_website_sets).

3. Permission is granted or denied based on whether the content meets all the security requirements — see [Security considerations](#security_considerations) for general requirements, and [Browser-specific variations](#browser-specific_variations) for some browser-specific security requirements.
   The {{jsxref("Promise")}}-based nature of `requestStorageAccess()` allows you to run code to handle success and failure cases.

   Once permission is granted, a permission key is stored in the browser with the structure `<top-level site, embedded site>`.
   For example, if the embedding site is `embedder.com`, and the embed is `locator.example.com`, the key would be `<embedder.com, example.com>`.

>    What this means is that permission is granted for access to unpartitioned cookies for any page on the `example.com` site or one of its subdomains that is embedded in any page on the `embedder.com` site.

Reviewer: ```suggestion
   This means that permission is granted for unpartitioned cookie access to any page on the `example.com` site or any of its subdomains embedded in any page on the `embedder.com` site.
```

   In such cases, the storage access is carried over from the previous navigation.
   This allows the embedded resource to reload itself and gain access to its cookies.

   > [!NOTE]
   > In older spec versions, the access was _per-page_ (Safari is the only browser still using this model). When one embed received third-party cookie access via `requestStorageAccess()`, all other same-site embeds would automatically receive access.
   > This was not desirable behavior from a security standpoint — for example, if `shop.example.com` embedded `locator.users.com` to allow users to use their location info while shopping, and `locator.users.com` called `requestStorageAccess()`, `shop.example.com` and any other sites it embeds would be able to access its cookies, but also access cookies from `private.users.com`, which is not intended to be embedded. [Read more about the motivations](https://github.com/privacycg/storage-access/issues/113) behind this change.

5. After an embed has activated the storage-access permission it should reload itself.

>    The browser will the re-request the resource with third party cookies included, and make them available to the embedded resource once it has loaded.

Reviewer: ```suggestion
   The browser will re-request the resource with third-party cookies included, and make them available to the embedded resource once it has loaded.
```

The workflows are shown in the [Storage access header sequences](#storage_access_header_sequences) section.

### Request/response flow

#### JavaScript sequences

Consider the example of a library to be loaded in an {{htmlelement("iframe")}} that needs to be shared among a number of sites, and which relies on credentials stored in unpartitioned cookies.

First consider the case where permission has not been granted.

- Resources are requested without third party cookies by default, so the server returns a version that does not rely on credentials (and which when loaded doesn't have access to the cookies).
- Once loaded, the resource calls `requestStorageAccess()` with transient activation to request and activate the `storage-access` permission.
- If permission is granted, the resource will then reload itself.
  This time the browser makes the request with third party cookies included, and the server may respond with a different "credentialed" version of the resource.
- The browser then loads the resource, and because it has an activated `storage-access` permission, gives it access to its cookies.

> ![Storage API workflow - without storage-access permission](storage_api_no_permission.png)

Reviewer: A couple of thoughts on the diagrams:

1. In each case, how about making the written steps numbered rather than bulleted, and adding numbers to the diagrams that match up to the numbered steps.
2. Watch your consistency in the diagrams. In some cases, you've got a space between the item text and the square-bracketed content, and in some cases, not. You've also got a full stop at the end of one diagram item, and no others. Also, in the third diagram, shouldn't the first request have "Request resource" above the header, for consistency?

Author: @chrisdavidmills Good suggestion. The bullets in the images somewhat force discipline in the docs. Done.

sequenceDiagram;
    participant Client
    participant Server
    Client->>Server: Request resource (no cookies)
    Server-- >>Client: Response[fallback content]
    Note over Client: Embed calls requestStorageAccess() with<br>transient activation to request permission
    Note over Client: User grants permission<br>Embed reloads itself (request includes cookies)
    Client->>Server: Request<br>Cookie: userid=123
    Server-- >>Client: Response [content]
    Note over Client: Client loads widget with SAA permission active

> -->

Author: FYI I've left these here to make it easy to edit the diagrams. I tried downloading the SVG from mermaid.ink but they aren't rendering.

Anyway, when done, I plan to remove the generated image but leave the sequence info.

Now consider the case where permission has been granted but not activated.
This would happen if you were to open the same URL in a new browser tab, or attempt to embed the same resource from another page in the same site.

The workflow is almost exactly the same because the resource still needs to be loaded the first time without cookies, and it then needs to call `requestStorageAccess()` to activate the permission for the context.

> In this case though it doesn't need transient activation and can execute on load.

Reviewer: ```suggestion
In this case, however, it doesn't need transient activation and can execute on load.
```

sequenceDiagram;
    participant Client
    participant Server
    Client->>Server: Request resource (no cookies)
    Server-- >>Client: Response[fallback content]
    Note over Client: Embed calls requestStorageAccess() to activate permission
    Note over Client: Embed reloads itself (request includes cookies)
    Client->>Server: Request<br>Cookie: userid=123
    Server-- >>Client: Response [content]
    Note over Client: Client loads widget with SAA permission active
-->

#### Storage access header sequences

The storage access headers enable an improved workflow that allows the server to request that the browser activate a permission that has been granted and retry the request with cookies included.

> This avoids having to load the resource, in order to call `requestStorageAccess()`, when the user has already granted permission.

Reviewer: ```suggestion
This avoids the requirement to load the resource to call `requestStorageAccess()` when the user has already granted permission.
```

The {{HTTPHeader("Sec-Fetch-Storage-Access")}} header is added to requests to indicate the storage access state of the current fetch context, such as whether permission has been activated, granted, or not granted.

> Depending on the storage access state of the request, the server can respond with an {{HTTPHeader("Activate-Storage-Access")}} to request that the browser activate the permission for the context and retry the request with cookies (avoiding it having to load the resource so it can call `requestStorageAccess()`).

Reviewer: ```suggestion
Depending on the storage access state of the request, the server can respond with an {{HTTPHeader("Activate-Storage-Access")}} header to request that the browser activate the permission for the context and retry the request with cookies.
```

You already said that above?

-->

What differs in the above sequence is that the browser and server have added some storage access headers (which aren't strictly needed).

- Initially the browser adds the header `Sec-Fetch-Storage-Access: none` to the request to indicate that permission has not been granted.
- After the user has granted (and thereby activated) the permission, the embed reloads itself.
  - The browser adds `Sec-Fetch-Storage-Access: active` to the request to indicate the context has an activated `storage-access` permission, and includes the third party cookies.
  - The server responds with `Activate-Storage-Access: load` which tells the browser to load the new version of the library with access to third party cookies.

The real benefit of the headers comes when attempting to load an embedded resource that already has permission granted.
In this case the browser will attach `Sec-Fetch-Storage-Access: inactive` to the request to indicate that the context has permission but that it is inactive.
The server can then respond with `Activate-Storage-Access: retry` to indicate that the browser should activate the permission, and retry the request with cookies.

> If the browser retries the request it adds `Sec-Fetch-Storage-Access: active` to the request along with the cookies, and the server responds with `Activate-Storage-Access: load` which tells the browser to load the new version of the library with access to third party cookies.

Reviewer: ```suggestion
If the browser retries the request, it adds `Sec-Fetch-Storage-Access: active` to the request along with the cookies. The server then responds with `Activate-Storage-Access: load`, which tells the browser to load the new version of the library with access to third-party cookies.
```
